### PR TITLE
Rename src/backend to src/dawn_native

### PR DESCRIPTION
**Pull request by Kangz**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -210,7 +210,7 @@ target_link_libraries(dawncpp dawn)
 add_subdirectory(third_party)
 
 add_subdirectory(src/common)
-add_subdirectory(src/backend)
+add_subdirectory(src/libDawn)
```
[the remark this answers is not in the document]
Comment (SenorBlanco) [reply CMakeLists.txt]: Just so I understand the vision here, what's the plan for the other dirs?

[the remark this answers is not in the document]
Comment (kainino0x) [reply CMakeLists.txt]: It mirrors `libANGLE` in ANGLE. However I'm not strongly attached to this rename myself.

[the remark this answers is not in the document]
Comment (Kangz) [reply CMakeLists.txt]: Thanks for the bikeshed, maybe we can improve on the ANGLE model. The dependencies of libDawn would be:
 - src/common
 - Some autogenerated files like the C/C++ headers, autogenerated validation etc.
 - SPIR-V Tools stuff like the validator
 - Our SPIR-V cross replacement.

The wire would be in a separate library, maybe libDawnWire because it doesn't need to know about internal details of libDawn. Tests and examples just like against libDawn.

Alternative names could be "implementation", "impl", "dawn", "main". I'd rather not split the frontend and backend in different directories because the nesting of backend namespaces in the frontend's namespace is very useful.

Or maybe something like this:
 - src/
   - common/
   - backend/ maybe with some backend-agnostic utility classes in namespace "dawn_impl::backend"
     - vulkan/ with namespace "dawn_impl::backend::vulkan"
     - other backends
  - frontend/ with namespace "dawn_impl"

@vonture might be able to share what he thinks good have worked better for ANGLE.

**Comment by vonture** [comment]
For ANGLE's case, ANGLE is simply the name of the implementation, our final exports are the libGLESv2 and EGL dlls so I wasn't worried about people trying to link libANGLE directly.  Are you able to draw the same parallel here or is dawn it's own API and dll that happens to be capable of implementing WebGPU?

When writing a shared library project, I do think that it makes sense to have most of the code in a static library and only implement the entry points in the shared library.  This forces you to not have any global state in most of your code and makes unittests much easier.  It looks like this is already the case?

**Comment by Kangz** [comment]
Thanks for the insights. The situation for Dawn / WebGPU is a bit different: WebGPU is only defined as a Javascript API and there is no concrete plan to spec a C or C++ header. So Dawn is just a C/C++ library that implements an API that's really close to WebGPU.

Currently the public API of Dawn is loosely defined but is mostly datatype definitions, a function pointer table structure ("proctable"), and some backend-specific APIs to create devices and proctables.

For now we don't do the proper exporting of symbols so libDawn will be a shared library. Eventually libDawn will be a source_set and will be shared by libDawn_static and libDawn a bit like ANGLE I think.

Basically I'm still not sure what we should do :)

**Comment by Kangz** [comment]
After offline discussion with @vonture we ended with the following. In term of libraries there is:
 - libdawn-native.so: that corresponds a bit to EGL or glfw_native.h and is basically what's currently in src/backend. It allows creating devices and proc-tables as well as doing backend-specific stuff.
 - libdawn.so: these are the dawn.c and dawncpp.cpp files essentially, helper libraries that can take a proctable and make it global.
  - libdawn-ipc.so or libdawn-wire.so: the ipc implementation.

None of these libraries depend on each-other but they all use the "dawn.h" header to share datatypes.

Because dawn-native is the main part of the implementation of dawn, it lives in src/dawn instead of src/backend. The last problem is that src/dawn would need to use the dawn namespace which would conflict with the dawncpp.h header. Mmmmh

**Comment by Kangz** [comment]
PTAL again, renamed to dawn_native instead.
